### ISSUE TITLE
Making Open3D classes pickleable

### ISSUE THREAD

**Issue by qianyizh** [isl-org/Open3D#218]
It is always good to have a convenient way to dump/load class instances to/from files.

This is doable with pybind11.
http://pybind11.readthedocs.io/en/stable/advanced/classes.html#pickling-support

I don't think this is high-priority. But it would be nice if it can be fixed in the future.

**Comment by tartavull** [comment]
+1

**Comment by HimangiM** [comment]
+1

**Comment by VinceDT** [comment]
+1

**Comment by tcies** [comment]
+1, working around by using a different format per default and converting to open3d only when needed for algorithms

**Comment by mberglundmx** [comment]
I needed TriangleMesh tobe pickable for my project, I managed it by just adding these lines to trianglemesh.cpp in Open3D\src\Python\open3d_pybind\geometry\trianglemesh.cpp:

```
.def(py::pickle([](const geometry::TriangleMesh &p) {  // _getstate_
        /* Return a tuple that fully encodes the state of the
        * object */

        py::object np = py::module::import("numpy");
        py::object vertices = np.attr("asarray")(p.vertices_);
        py::object triangles = np.attr("asarray")(p.triangles_);
        py::object triangle_normals = np.attr("asarray")(p.triangle_normals_);
    
        return py::make_tuple(vertices, triangles, triangle_normals);
    }, [](py::tuple t) {  // _setstate_
        if (t.size() != 3) throw std::runtime_error("Invalid state!");
        
        /* Create a new C++ instance */
        geometry::TriangleMesh p = geometry::TriangleMesh();
        py::object o3d = py::module::import("open3d");
        
        p.vertices_ = o3d.attr("utility").attr("Vector3dVector")(t[0]).cast<std::vector<Eigen::Vector3d>>();
        p.triangles_ = o3d.attr("utility").attr("Vector3iVector")(t[1]).cast<std::vector<Eigen::Vector3i>>();
        p.triangle_normals_ = o3d.attr("utility").attr("Vector3dVector")(t[2]).cast<std::vector<Eigen::Vector3d>>();

        return p;
}))
```

Not the most efficient solution, but it solves my problem where I need to be able to archive the mesh data.

**Comment by germanros1987** [comment]
@griegler do you have an opinion about this? Do you think we should make TriangleMesh pickable?

**Comment by griegler** [comment]
I think we should prioritize this for the new geometry classes. @yxlao

**Comment by IvanCobotic** [comment]
+1
I often have need to pickle various intermediate results and converting to/from Open3D format introduces a lot of unnecessary code.

**Comment by sandeepnmenon** [comment]
+1
I get pickling error while trying to use multiprocessing with open3d objects
`TypeError: can't pickle open3d.cpu.pybind.geometry.PointCloud objects
`

**Comment by cansik** [comment]
I am looking for a way to use multi-processing to speed up the loading of triangle meshes.
Is there already a method to quickly load a lot of meshes or how would you suggest doing it without being able to pickle the objects?

**Comment by cansik** [comment]
A simple workaround would be to create a decorator object that implements pickleable types for the attributes. I have done something (a quick draft) for a project where I need to multiprocess the mesh loading: [FastGeometryLoader.py#L10-L50](https://github.com/cansik/mesh-sequence-player/blob/main/mesh_sequence_player/FastGeometryLoader.py#L10-L50)

So in general what I did was to just convert all the necessary attributes into numpy arrays and than convert them back into the Eigen datatypes. For a next iteration I would maybe create a decorator object where the `__reduce__` method is implemented to do this. But for now it's fine.

**Comment by nachovizzo** [comment]
@cansik++ thanks for sharing the code.

For those who need a more "simple" functionality you could also do something like
```python
import numpy as np
import open3d as o3d


class SerializableMesh:
    def __init__(self, mesh: o3d.geometry.TriangleMesh):
        self.vertices = np.asarray(mesh.vertices)
        self.triangles = np.asarray(mesh.triangles)

    def to_open3d(self) -> o3d.geometry.TriangleMesh:
        return o3d.geometry.TriangleMesh(
            vertices=o3d.utility.Vector3dVector(self.vertices),
            triangles=o3d.utility.Vector3iVector(self.triangles),
        )
```

Of course, you don't store all mesh properties, but it's a starting point

**Comment by macawi** [comment]
+1
Performing conversion to numpy arrays and back to o3d object just to pass it through multiprocessing.Pool.map.

**Comment by RMeli** [comment]
+1
As already pointed out by other people, I tired to use [mpire](https://github.com/Slimmer-AI/mpire) to process multiple point clouds in parallel, but got the following error:
```
TypeError: can't pickle open3d.cpu.pybind.geometry.PointCloud objects
```
Having this feature would be extremely useful.

**Comment by cansik** [comment]
@RMeli If you need a similar pickable object for PointClouds, you can use this:

```python
class _PointCloudTransmissionFormat:
    def __init__(self, pointcloud: PointCloud):
        self.points = np.array(pointcloud.points)
        self.colors = np.array(pointcloud.colors)
        self.normals = np.array(pointcloud.normals)

    def create_pointcloud(self) -> PointCloud:
        pointcloud = PointCloud()
        pointcloud.points = o3d.utility.Vector3dVector(self.points)
        pointcloud.colors = o3d.utility.Vector3dVector(self.colors)
        pointcloud.normals = o3d.utility.Vector3dVector(self.normals)
        return pointcloud
```

And here is an example on how to load PointClouds in parallel:

```python
def _load_pointcloud_data(file: str) -> _PointCloudTransmissionFormat:
    pointcloud = o3d.io.read_point_cloud(file)
    return _PointCloudTransmissionFormat(pointcloud)

def load_pointclouds_fast(files: [str]) -> [PointCloud]:
    clouds = []
    with Pool(processes=multiprocessing.cpu_count()) as pool:
        for result in tqdm(pool.imap(_load_pointcloud_data, files), total=len(files), desc="pointcloud loading"):
            clouds.append(result)
    return [cloud.create_pointcloud() for cloud in clouds]
```

**Comment by RMeli** [comment]
Thanks @cansik , I'll try to adapt this for my needs.

**Comment by HOUYONGKUO** [comment]
TypeError: cannot pickle 'open3d.cpu.pybind.utility.Vector3dVector' object

**Comment by pauldeee** [comment]
+1
I just ran into this issue trying to use Pool. @cansik thanks, your solution works well.

**Comment by Guptajakala** [comment]
+1

**Comment by wambomeister** [comment]
+1

**Comment by ssheorey** [comment]
Pickling support has been available for the new Tensor APi geometry classes (including point cloud, mesh, etc.) since v0.16. Please see here:
http://www.open3d.org/docs/latest/tutorial/core/tensor.html?highlight=pickle#Pickle-support

If your code still uses older Eigen based classes, you can use the `from_legacy()` / `to_legacy()` functions for conversion before / after saving.

**Comment by JohannesLengler** [comment]
The new tensor based OrientedBoundingBox doesn't seem to support pickling on 0.17.0. Is this something one could expect to work in the future? For now I'm using `get_box_points` and `create_from_points` to pickle the bbox as an intermediate array, but native support would be nice to have

**Comment by louhz** [comment]
does tetramesh supported? looks like it raise error:TypeError: cannot pickle 'open3d.cuda.pybind.geometry.TetraMesh' object
I guess this is because the tetrahedra contain ordered set of vertices node?

**Comment by VladimirYugay** [comment]
Seems like PoseGraph is not supported?

**Comment by ssheorey** [comment]
Hi @VladimirYugay that's correct. Please open a new issue for the specific classes that you are looking for. PRs adding pickling support for these classes are also welcome!

**Comment by jtzhangLogan** [comment]
Is it possible to add pickle support for o3d.geometry.KDTreeFlann?
